### PR TITLE
Add OTLP for Activity Exporter 

### PR DESCRIPTION
**Pull request by pjanotti**
Per our conversation, this is the Activity Exporter for OTLP consuming `Activity` batches. 

I opted at this time to keep `ActivityExporter` as it is (there was a quick chat on Gitter on perhaps adding the `ExportAsync(IEnumerable<Activity>, CancellationToken)` but let's have a separate discussion for it).

I will be commenting on some points of the PR to some of the potential issues/follow-ups.

### Checklist
- [X] I ran Unit Tests locally.

### CODE REVIEW

File: samples/Exporters/Console/TestOtlp.cs [hunk 1]
```diff
@@ -47,5 +57,28 @@ internal static object Run(string endpoint)
...
             return null;
         }
+
+        private static object RunWithActivitySource(string endpoint)
+        {
+            // Enable OpenTelemetry for the source "MyCompany.MyProduct.MyWebServer"
+            // and use Console exporter
+            OpenTelemetrySdk.EnableOpenTelemetry(
+                builder => builder
+                    .AddActivitySource("Samples.SampleServer")
```
Comment (pjanotti) [inline samples/Exporters/Console/TestOtlp.cs:67]: @cijothomas I'm leaning into listening to all sources by default and filter out un-wanted ones. Depending on how some libraries are instrumented we may need some processors to filter out some activities.

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ActivityExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+            {
+                Name = activity.OperationName, // TODO: map DisplayName to anything?
+
+                Kind = (OtlpTrace.Span.Types.SpanKind)(activity.Kind + 1), // TODO: there is an offset of 1 on the enum.
+
+                TraceId = ByteString.CopyFrom(traceIdBytes.ToArray()),
+                SpanId = ByteString.CopyFrom(spanIdBytes.ToArray()),
+                ParentSpanId = parentSpanIdString,
+
+                // TODO: Status is still pending, need to pursue OTEL spec change.
```
Comment (pjanotti) [inline src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ActivityExtensions.cs:102]: Following up on https://github.com/open-telemetry/opentelemetry-specification/issues/306#issuecomment-628961881

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ActivityExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+            {
+                Span<byte> parentSpanIdBytes = stackalloc byte[8];
+                activity.ParentSpanId.CopyTo(parentSpanIdBytes);
+                parentSpanIdString = ByteString.CopyFrom(parentSpanIdBytes.ToArray());
+            }
+
+            var startTimeUnixNano = activity.StartTimeUtc.ToUnixTimeNanoseconds();
+            var otlpSpan = new OtlpTrace.Span
+            {
+                Name = activity.OperationName, // TODO: map DisplayName to anything?
```
Comment (pjanotti) [inline src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ActivityExtensions.cs:94]: In principle, I'm assuming that we are not going to use `DisplayName`.

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ActivityExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+            };
+
+            foreach (var kvp in activity.Tags)
+            {
+                // TODO: attempt to convert to supported types?
+                // TODO: enforce no duplicate keys?
+                // TODO: drop if Value is null?
+                // TODO: reverse?
+                var attrib = new OtlpCommon.AttributeKeyValue { Key = kvp.Key, StringValue = kvp.Value };
+                otlpSpan.Attributes.Add(attrib);
```
Comment (pjanotti) [inline src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ActivityExtensions.cs:115]: @reyang @cijothomas @noahfalk I think that this is one of the big things to have a wrapper so far: here OTel defines some behaviors that are quite different. In principle, we could have a processor to enforce the OTel spec here but that takes away the control from authors that want to enforce that spec on their code. The OTel related spec: https://github.com/open-telemetry/opentelemetry-specification/blob/bdf39cb935a6f48badc687daeb78a66dc352761f/specification/trace/api.md#set-attributes

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ActivityExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+
+            // Activity does not limit number of attributes, events, links, etc so drop counts are always zero.
+
+            return otlpSpan;
+        }
+
+        private static Dictionary<Resource, Dictionary<ActivitySource, List<OtlpTrace.Span>>> GroupByResourceAndLibrary(
+            IEnumerable<Activity> activityBatch)
+        {
+            // TODO: there is no Resource associated here, other SDKs associate it to the span, that's not an option here.
```
Comment (pjanotti) [inline src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ActivityExtensions.cs:129]: There is already an issue tracking this, see #649.

File: samples/Exporters/Console/TestOtlp.cs [hunk 6]
```diff
@@ -47,5 +57,28 @@ internal static object Run(string endpoint)
 
             return null;
         }
+
+        private static object RunWithActivitySource(string endpoint)
+        {
+            // Enable OpenTelemetry for the source "MyCompany.MyProduct.MyWebServer"
```
Comment (cijothomas) [inline samples/Exporters/Console/TestOtlp.cs:63]: nit: wrong sourcename

File: samples/Exporters/Console/TestOtlp.cs [hunk 7]
```diff
@@ -47,5 +57,28 @@ internal static object Run(string endpoint)
...
+            // and use Console exporter
```
Comment (cijothomas) [inline samples/Exporters/Console/TestOtlp.cs:64]: change comment to OTLP exporter

File: samples/Exporters/Console/TestOtlp.cs (hunk 1 above)
Comment (cijothomas) [reply samples/Exporters/Console/TestOtlp.cs:67]: Yup. Lets address this separately.
Libraries like httpclient, which was instrumented using DiagnosticSource, will have their activities emitted with empty ActivitySource. Listening to them requires `AddActivitySource("")` now. 
Anyway, lets address this after some exporters/adapters are ready with Activity API.

File: samples/Exporters/Console/InstrumentationWithActivitySource.cs [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+
+                            using var activity = source.StartActivity(
+                                $"{context.Request.HttpMethod}:{context.Request.Url.AbsolutePath}",
+                                ActivityKind.Server);
+
+                            string requestContent;
+                            using (var reader = new StreamReader(context.Request.InputStream, context.Request.ContentEncoding))
+                            {
+                                var headerKeys = context.Request.Headers.AllKeys;
+                                foreach (var headerKey in headerKeys)
```
Comment (cijothomas) [inline samples/Exporters/Console/InstrumentationWithActivitySource.cs:76]: this header loop can be outside the using(reader) right?

File: samples/Exporters/Console/InstrumentationWithActivitySource.cs [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+                        while (!cancellationToken.IsCancellationRequested)
+                        {
+                            var content = new StringContent($"client message: {DateTime.Now}", Encoding.UTF8);
+
+                            using (var activity = source.StartActivity("POST:" + RequestPath, ActivityKind.Client))
+                            {
+                                activity.AddBaggage("client.request.count", count.ToString(CultureInfo.InvariantCulture));
+                                count++;
+
+                                activity.AddEvent(new ActivityEvent("PostAsync:Started"));
```
Comment (cijothomas) [inline samples/Exporters/Console/InstrumentationWithActivitySource.cs:132]: activity null checks needed everywhere

File: samples/Exporters/Console/InstrumentationWithActivitySource.cs [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+                        using var client = new HttpClient();
+
+                        var count = 1;
+                        while (!cancellationToken.IsCancellationRequested)
+                        {
+                            var content = new StringContent($"client message: {DateTime.Now}", Encoding.UTF8);
+
+                            using (var activity = source.StartActivity("POST:" + RequestPath, ActivityKind.Client))
+                            {
+                                activity.AddBaggage("client.request.count", count.ToString(CultureInfo.InvariantCulture));
```
Comment (cijothomas) [inline samples/Exporters/Console/InstrumentationWithActivitySource.cs:129]: did you intentionally use Baggage? Baggage is meant for properties which require to be transported to the next layer. Unless there is a reason to use it, lets not use it in the example.

File: samples/Exporters/Console/InstrumentationWithActivitySource.cs [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+using System.IO;
+using System.Net;
+using System.Net.Http;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Samples
+{
+    internal class InstrumentationWithActivitySource : IDisposable
```
Comment (cijothomas) [inline samples/Exporters/Console/InstrumentationWithActivitySource.cs:30]: this client-server example is very useful. we should modify other examples to use this one than simple cases. (not for this PR, but adding a note)

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/ActivitySourceExporter.cs [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+
+using OpenTelemetry.Exporter.OpenTelemetryProtocol.Implementation;
+using OpenTelemetry.Trace.Export;
+
+namespace OpenTelemetry.Exporter.OpenTelemetryProtocol
+{
+    /// <summary>
+    /// Exporter consuming data from <see cref="ActivitySource"/> sending OpenTelemetry protocol (OTLP).
+    /// </summary>
+    public class ActivitySourceExporter : ActivityExporter
```
Comment (cijothomas) [inline src/OpenTelemetry.Exporter.OpenTelemetryProtocol/ActivitySourceExporter.cs:30]: suggesttion to rename ActivitySourceExporter -> OtlpActivityExporter

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/ActivitySourceExporter.cs [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+
+using OpenTelemetry.Exporter.OpenTelemetryProtocol.Implementation;
+using OpenTelemetry.Trace.Export;
+
+namespace OpenTelemetry.Exporter.OpenTelemetryProtocol
+{
+    /// <summary>
+    /// Exporter consuming data from <see cref="ActivitySource"/> sending OpenTelemetry protocol (OTLP).
```
Comment (cijothomas) [inline src/OpenTelemetry.Exporter.OpenTelemetryProtocol/ActivitySourceExporter.cs:28]: perhaps simply state "Exporter exporting Activity data". Activities can come from ActivitySource, but they can come without ActivitySource as well.

File: samples/Exporters/Console/InstrumentationWithActivitySource.cs (hunk 10 above)
Comment (pjanotti) [reply samples/Exporters/Console/InstrumentationWithActivitySource.cs:129]: I was exploring the API and forgot to remove it.

File: samples/Exporters/Console/InstrumentationWithActivitySource.cs (hunk 9 above)
Comment (pjanotti) [reply samples/Exporters/Console/InstrumentationWithActivitySource.cs:132]: ops

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ActivityExtensions.cs [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+            return resourceSpansList;
+        }
+
+        internal static OtlpTrace.Span ToOtlpSpan(this Activity activity)
+        {
+            // protobuf doesn't understand Span<T> yet: https://github.com/protocolbuffers/protobuf/issues/3431
+            Span<byte> traceIdBytes = stackalloc byte[16];
+            Span<byte> spanIdBytes = stackalloc byte[8];
+
+            activity.TraceId.CopyTo(traceIdBytes);
```
Comment (cijothomas) [inline src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ActivityExtensions.cs:80]: i think its safer to check if activity.IDFormat == W3C. OT SDK sets it to W3C, but the exporter could be used independently?

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ActivityExtensions.cs (hunk 3 above)
Comment (cijothomas) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ActivityExtensions.cs:94]: I think the goal is to use DisplayName everywhere. OperationName was part of Activity already and it was not "settable" after construction. The SpanName in OT needs to be settable afterwards, which is why Activity introduced DisplayName.

In short, always use DisplayName and forget about OperationName. (DisplayName gets the OperationName by default. DisplayName can be changed, but OperationName cannot be)

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ActivityExtensions.cs (hunk 2 above)
Comment (cijothomas) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ActivityExtensions.cs:102]: This is planned to be discussed in OT Spec SIG tomorrow.
https://docs.google.com/document/d/1-bCYkN-DWJq4jw1ybaDZYYmx-WAe6HnwfWbkm8d57v8/edit#heading=h.evt7z82bt2gy

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ActivityExtensions.cs (hunk 4 above)
Comment (cijothomas) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ActivityExtensions.cs:115]: This is being addressed in Tarek and Noah. We can live with this now, and hopefully get better solution in next preview of DiagnosticSource.

Comment (cijothomas) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ActivityExtensions.cs:115]: Agree this is a valid reason for shim/wrapper to enforce things like no duplicates etc.

File: src/OpenTelemetry.Exporter.Console/ConsoleActivityExporter.cs [hunk 15]
```diff
@@ -79,6 +80,19 @@ public override Task<ExportResult> ExportAsync(IEnumerable<Activity> activityBat
                         foreach (var activityEvent in activity.Events)
                         {
                             System.Console.WriteLine($"Event Name: {activityEvent.Name} TimeStamp: {activityEvent.Timestamp}");
+                            foreach (var attribute in activityEvent.Attributes)
```
Comment (cijothomas) [inline src/OpenTelemetry.Exporter.Console/ConsoleActivityExporter.cs:83]: Thanks!

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ActivityExtensions.cs (hunk 14 above)
Comment (pjanotti) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ActivityExtensions.cs:80]: I will add the defensive code but I don't see much value in offering the exporter without the SDK.

Comment (cijothomas) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ActivityExtensions.cs:80]: agree.

**Pull request merged**